### PR TITLE
Make `HivePartitionSensorAsync` example dag self-sufficient.

### LINKED ISSUE #265: Fix HivePartitionSensor example DAG
- Change the docker file to copy the hive,  hadoop binaries and JDK 8 to airflow worker
- Deploy on staging and test the DAG end to end

### PR DESCRIPTION
**Pull request by rajaths010494**
- Make example dag of `HivePartitionSensorAsync` self-sufficient for integration testing.
- Update sphinx docs and doc-strings to `HivePartitionSensorAsync`.
- The library version of hive and Hadoop in ``Dockerfile`` should match the remote cluster where the
     Hadoop and hive are running.

Closes #265

### CODE REVIEW

File: .circleci/integration-tests/Dockerfile [hunk 1]
```diff
@@ -28,12 +38,28 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
     && chmod +x ./kubectl \
     && mv ./kubectl /usr/local/bin
 
+#install Apache Hive 2.3.9 (Download same version of library to match EMR hive version created by DAG)
+RUN curl "https://downloads.apache.org/hive/hive-2.3.9/apache-hive-2.3.9-bin.tar.gz" -o  "/tmp/apache-hive-2.3.9-bin.tar.gz"  \
```
Comment (phanikumv) [inline .circleci/integration-tests/Dockerfile:42]: consider reading the hive and hadoop versions from environment variables than hard coding the version

File: .circleci/integration-tests/Dockerfile [hunk 2]
```diff
@@ -28,12 +38,28 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
...
+    && tar -xf /tmp/apache-hive-2.3.9-bin.tar.gz -C /usr/local
+ENV HIVE_HOME=/usr/local/apache-hive-2.3.9-bin
+
+#install Apache Hadoop 2.10.1 (Download same version of library to match EMR hadopp version created by DAG)
+RUN curl "https://archive.apache.org/dist/hadoop/common/hadoop-2.10.1/hadoop-2.10.1.tar.gz" -o "/tmp/hadoop-2.10.1.tar.gz" \
+    && tar -xf /tmp/hadoop-2.10.1.tar.gz -C /usr/local
+ENV HADOOP_HOME=/usr/local/hadoop-2.10.1
+ENV HADOOP_CONF_DIR=$HADOOP_HOME/conf
+
+RUN curl "https://raw.githubusercontent.com/apache/hive/master/data/conf/hive-site.xml" -o "/tmp/hive-site.xml" \
```
Comment (phanikumv) [inline .circleci/integration-tests/Dockerfile:52]: why do we need this, if we are not using EMR version of hive-site.xml?

File: .circleci/integration-tests/Dockerfile [hunk 3]
```diff
@@ -71,6 +107,22 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
     && chmod +x ./kubectl \
     && mv ./kubectl /usr/local/bin
 
+#install Apache Hive 2.3.9 (Download same version of library to match EMR hive version created by DAG)
+RUN curl "https://downloads.apache.org/hive/hive-2.3.9/apache-hive-2.3.9-bin.tar.gz" -o  "/tmp/apache-hive-2.3.9-bin.tar.gz"  \
```
Comment (phanikumv) [inline .circleci/integration-tests/Dockerfile:111]: consider reading the hive and hadoop versions from environment variables than hard coding the version

File: astronomer/providers/apache/hive/example_dags/example_hive.py [hunk 4]
```diff
@@ -42,6 +42,9 @@
     "hdfs dfs -put zipcodes.csv /user/root",
 ]
 
+# emr-5.34.0 cluster has been used and apache-hive-2.3.9-bin and hadoop-2.10.1
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/example_dags/example_hive.py:45]:
```suggestion
# This example uses an emr-5.34.0 cluster, apache-hive-2.3.9 and hadoop-2.10.1
```

File: astronomer/providers/apache/hive/example_dags/example_hive.py [hunk 5]
```diff
@@ -42,6 +42,9 @@
...
+# has been used for integration testing. If the versions of emr cluster changes then we need to
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/example_dags/example_hive.py:46]:
```suggestion
# If you would like to use a different version of the EMR cluster, then we need to
```

File: astronomer/providers/apache/hive/example_dags/example_hive.py [hunk 6]
```diff
@@ -42,6 +42,9 @@
...
+# update and match the hive and hadoop libraries versions in the integration tests `Dockefile`.
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/example_dags/example_hive.py:47]:
```suggestion
# match the hive and hadoop versions same as specified in the integration tests `Dockefile`.
```

**Review by phanikumv — CHANGES_REQUESTED** [review]

File: astronomer/providers/apache/hive/sensors/hive_partition.py [hunk 7]
```diff
@@ -12,9 +12,11 @@ class HivePartitionSensorAsync(HivePartitionSensor):
     """
     Waits for a given partition to show up in Hive.
     Note: HivePartitionSensorAsync uses implya library instead of pyhive
-    since pyhive is currently unsupported. Refer https://github.com/dropbox/PyHive.
-    Since we use implya library, please set the connection to use the port 10000 instead of 9083
-    The sensor currently supports auth_mechansim='PLAIN' only.
+    since pyhive is currently `unsupported <https://github.com/dropbox/PyHive>`_.
```
Comment (pankajastro) [inline astronomer/providers/apache/hive/sensors/hive_partition.py:15]: you can write a note 
```
.. note::
    This function is not suitable for sending spam e-mails.
```

File: .circleci/integration-tests/Dockerfile (hunk 1 above)
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: Done added ENV variable

File: .circleci/integration-tests/Dockerfile (hunk 2 above)
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: removed not needed. It worked with out it

File: .circleci/integration-tests/Dockerfile (hunk 3 above)
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: Done added ENV variable

File: astronomer/providers/apache/hive/example_dags/example_hive.py (hunk 4 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/example_dags/example_hive.py]: done

File: astronomer/providers/apache/hive/example_dags/example_hive.py (hunk 5 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/example_dags/example_hive.py]: done

File: astronomer/providers/apache/hive/example_dags/example_hive.py (hunk 6 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/example_dags/example_hive.py]: done

File: astronomer/providers/apache/hive/sensors/hive_partition.py (hunk 7 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/sensors/hive_partition.py]: added .. note

File: .circleci/integration-tests/Dockerfile [hunk 8]
```diff
@@ -71,6 +112,19 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
     && chmod +x ./kubectl \
     && mv ./kubectl /usr/local/bin
 
+#install Apache Hive (Download same version of library to match EMR hive version created by DAG)
+RUN curl "https://downloads.apache.org/hive/$HIVE_LIBRARY_VERSION/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz" -o  "/tmp/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz"  \
+    && tar -xf /tmp/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz -C /usr/local
+ENV HIVE_HOME=/usr/local/apache-$HIVE_LIBRARY_VERSION-bin
+
+#install Apache Hadoop (Download same version of library to match EMR hadopp version created by DAG)
```
Comment (phanikumv) [inline .circleci/integration-tests/Dockerfile:120]: correct the typo
```suggestion
#install Apache Hadoop (Download same version of library to match EMR hadoop version created by DAG)
```

File: astronomer/providers/apache/hive/example_dags/example_hive.py [hunk 9]
```diff
@@ -162,7 +165,7 @@ def add_inbound_rule_for_security_group(task_instance: Any) -> None:
                 ip_exists = True
 
     if not ip_exists:
-        # open port for port 8998
+        # open port for port 10000
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/example_dags/example_hive.py:168]:
```suggestion
        # Port 10000 need to be open for Impyla connectivity to Hive
```

File: astronomer/providers/apache/hive/example_dags/example_hive.py [hunk 10]
```diff
@@ -176,6 +179,7 @@ def add_inbound_rule_for_security_group(task_instance: Any) -> None:
                 }
             ],
         )
+
         # open port for port 22 for ssh and copy file for hdfs
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/example_dags/example_hive.py:183]:
```suggestion
        # Allow SSH traffic on port 22 and copy file to hdfs
```

File: astronomer/providers/apache/hive/sensors/hive_partition.py [hunk 11]
```diff
@@ -10,11 +10,16 @@
...
+    .. note:: HivePartitionSensorAsync uses implya library instead of pyhive
+       since pyhive is currently `unsupported <https://github.com/dropbox/PyHive>`_.
+
+       Since we use implya library, please set the connection to use the port ``10000`` instead of ``9083``
+       The sensor currently supports ``auth_mechansim='PLAIN'`` only.
+
+       The library version of hive and hadoop in ``Dockerfile`` should match the remote cluster where the
+       hadoop and hive is running.
+
+    Waits for a given partition to show up in Hive table.
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/sensors/hive_partition.py:22]: can you check how the docs are being generated for this block

Comment (phanikumv) [inline astronomer/providers/apache/hive/sensors/hive_partition.py:22]: `Waits for a given partition to show up in Hive.` should come first in the docstring, and then you can mention the note

File: astronomer/providers/apache/hive/sensors/hive_partition.py [hunk 12]
```diff
@@ -10,11 +10,16 @@
 
 class HivePartitionSensorAsync(HivePartitionSensor):
     """
-    Waits for a given partition to show up in Hive.
-    Note: HivePartitionSensorAsync uses implya library instead of pyhive
-    since pyhive is currently unsupported. Refer https://github.com/dropbox/PyHive.
-    Since we use implya library, please set the connection to use the port 10000 instead of 9083
-    The sensor currently supports auth_mechansim='PLAIN' only.
+    .. note:: HivePartitionSensorAsync uses implya library instead of pyhive
+       since pyhive is currently `unsupported <https://github.com/dropbox/PyHive>`_.
```
Comment (phanikumv) [inline astronomer/providers/apache/hive/sensors/hive_partition.py:14]:
```suggestion
       Note that pyhive is currently `unsupported <https://github.com/dropbox/PyHive>`_.
```

**Review by phanikumv — CHANGES_REQUESTED** [review]

File: astronomer/providers/apache/hive/sensors/hive_partition.py (hunk 11 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/sensors/hive_partition.py:22]: image

File: astronomer/providers/apache/hive/example_dags/example_hive.py (hunk 9 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/example_dags/example_hive.py]: Done

File: .circleci/integration-tests/Dockerfile (hunk 8 above)
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: Done

File: astronomer/providers/apache/hive/example_dags/example_hive.py (hunk 10 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/example_dags/example_hive.py]: Done

File: astronomer/providers/apache/hive/sensors/hive_partition.py (hunk 11 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/sensors/hive_partition.py]: Moved to the beginning of docstring.
.

File: astronomer/providers/apache/hive/sensors/hive_partition.py (hunk 12 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/sensors/hive_partition.py]: Done

File: astronomer/providers/apache/hive/sensors/hive_partition.py (hunk 11 above)
Comment (phanikumv) [reply astronomer/providers/apache/hive/sensors/hive_partition.py]: Make the below changes please
- The sync version of this sensor uses `pyhive`, but `pyhive` is currently unsupported
- you can put a hyperlink on the `impyla` and `pyhive`
- a fullstop is missing after 9083
- This sensor currently..

**Review by phanikumv — CHANGES_REQUESTED** [review]
Please make the changes so that it shows up nicely on Sphinx,
rest of the code looks good

File: astronomer/providers/apache/hive/sensors/hive_partition.py (hunk 11 above)
Comment (rajaths010494) [reply astronomer/providers/apache/hive/sensors/hive_partition.py]: Done.
image

**Review by phanikumv — COMMENTED** [review]
I want a review on this from all CODEOWNERS please @pankajastro @pankajkoti @bharanidharan14 @rajaths010494 

cc @kaxil

File: astronomer/providers/apache/hive/example_dags/example_hive.py [hunk 13]
```diff
@@ -42,6 +42,9 @@
     "hdfs dfs -put zipcodes.csv /user/root",
 ]
 
+# This example uses an emr-5.34.0 cluster,apache-hive-2.3.9-bin and hadoop-2.10.1.
```
Comment (pankajastro) [inline astronomer/providers/apache/hive/example_dags/example_hive.py:45]:
```suggestion
# This example uses an emr-5.34.0 cluster, apache-hive-2.3.9-bin and hadoop-2.10.1.
```

**Comment by pankajastro** [comment]
LGTM

**Review by pankajastro — APPROVED** [review]

File: astronomer/providers/apache/hive/example_dags/example_hive.py (hunk 13 above)
Comment (phanikumv) [reply astronomer/providers/apache/hive/example_dags/example_hive.py:45]:
```suggestion
# This example uses an emr-5.34.0 cluster, apache-hive-2.3.9 and hadoop-2.10.1.
```

Comment (rajaths010494) [reply astronomer/providers/apache/hive/example_dags/example_hive.py]: Done

File: .circleci/integration-tests/Dockerfile [hunk 14]
```diff
@@ -71,6 +112,19 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
...
+    && tar -xf /tmp/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz -C /usr/local
+ENV HIVE_HOME=/usr/local/apache-$HIVE_LIBRARY_VERSION-bin
+
+#install Apache Hadoop (Download same version of library to match EMR hadoop version created by DAG)
+RUN curl "https://archive.apache.org/dist/hadoop/common/$HADOOP_LIBRARY_VERSION/$HADOOP_LIBRARY_VERSION.tar.gz" -o "/tmp/$HADOOP_LIBRARY_VERSION.tar.gz" \
+    && tar -xf /tmp/$HADOOP_LIBRARY_VERSION.tar.gz -C /usr/local
+ENV HADOOP_HOME=/usr/local/$HADOOP_LIBRARY_VERSION
+ENV HADOOP_CONF_DIR=$HADOOP_HOME/conf
+
+ENV PATH $PATH:$JAVA_HOME/bin:$HIVE_HOME:$HIVE_HOME/bin:$HADOOP_HOME:$HADOOP_HOME/bin
```
Comment (pankajkoti) [inline .circleci/integration-tests/Dockerfile:126]: Do we also need `$HIVE_HOME` and `$HADOOP_HOME` in the path. Mean to ask that isn't the paths for binaries (`$HIVE_HOME/bin` &  `HADOOP_HOME/bin`) enough?

File: .circleci/integration-tests/Dockerfile [hunk 15]
```diff
@@ -28,12 +42,25 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
...
```
Comment (pankajkoti) [inline .circleci/integration-tests/Dockerfile:56]: same question for paths.

**Review by pankajkoti — APPROVED** [review]
Looks good to me. Minor suggestions and questions.

File: .circleci/integration-tests/Dockerfile (hunk 15 above)
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile:56]: We only need executables set in the PATh `$HAOOP_HOME` and `$HIVE_HOME` can be removed from the path.

File: .circleci/integration-tests/Dockerfile [hunk 16]
```diff
@@ -71,6 +112,19 @@ RUN curl -o kubectl https://s3.us-west-2.amazonaws.com/amazon-eks/1.22.6/2022-03
     && chmod +x ./kubectl \
     && mv ./kubectl /usr/local/bin
 
+#install Apache Hive (Download same version of library to match EMR hive version created by DAG)
+RUN curl "https://downloads.apache.org/hive/$HIVE_LIBRARY_VERSION/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz" -o  "/tmp/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz"  \
+    && tar -xf /tmp/apache-$HIVE_LIBRARY_VERSION-bin.tar.gz -C /usr/local
+ENV HIVE_HOME=/usr/local/apache-$HIVE_LIBRARY_VERSION-bin
+
+#install Apache Hadoop (Download same version of library to match EMR hadoop version created by DAG)
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: Done

File: astronomer/providers/apache/hive/sensors/hive_partition.py [hunk 17]
```diff
@@ -10,11 +10,19 @@
...
+       HivePartitionSensorAsync uses implya library instead of pyhive.
+       The sync version of this sensor uses `pyhive <https://github.com/dropbox/PyHive>`_,
+       but `pyhive <https://github.com/dropbox/PyHive>`_ is currently unsupported.
+
+       Since we use `implya <https://github.com/cloudera/impyla>`_ library,
+       please set the connection to use the port ``10000`` instead of ``9083``.
+       This sensor currently supports ``auth_mechansim='PLAIN'`` only.
+
+       The library version of hive and hadoop in ``Dockerfile`` should match the remote cluster where the
+       hadoop and hive is running.
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply astronomer/providers/apache/hive/sensors/hive_partition.py]: Done.

File: .circleci/integration-tests/Dockerfile [hunk 18]
```diff
@@ -46,6 +73,16 @@ FROM quay.io/astronomer/astro-runtime:4.2.4-base as astro-cloud
...
 
 USER root
+
+RUN apt-get update -y \
+    && apt-get install -y software-properties-common \
+    && apt-get install -y wget procps gnupg2
+
+# Install openjdk-8
+RUN apt-add-repository 'deb http://security.debian.org/debian-security stretch/updates main'
+RUN apt-get update && apt-get install -y openjdk-8-jdk
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: Done.

File: .circleci/integration-tests/Dockerfile [hunk 19]
```diff
@@ -3,6 +3,16 @@ FROM quay.io/astronomer/ap-airflow:2.2.5 as staging
...
```
[the remark this answers is not in the document]
Comment (rajaths010494) [reply .circleci/integration-tests/Dockerfile]: Done.

**Review by phanikumv — APPROVED** [review]

**Pull request merged**